### DISCUSSION TITLE
Why does the SystemState interface accept World in `new` and then again in `get`/`get_mut`?

### DISCUSSION THREAD

**Discussion by MeoMix** [bevyengine/bevy#11724] [Q&A] [answered]
SystemState's code looks like this:

```rust

    pub fn new(world: &mut World) -> Self {
        let mut meta = SystemMeta::new::<Param>();
        meta.last_run = world.change_tick().relative_to(Tick::MAX);
        let param_state = Param::init_state(world, &mut meta);
        Self {
            meta,
            param_state,
            world_id: world.id(),
            archetype_generation: ArchetypeGeneration::initial(),
        }
    }

    /// Retrieve the [`SystemParam`] values. This can only be called when all parameters are read-only.
    #[inline]
    pub fn get<'w, 's>(&'s mut self, world: &'w World) -> SystemParamItem<'w, 's, Param>
    where
        Param: ReadOnlySystemParam,
    {
        self.validate_world(world.id());
        self.update_archetypes(world);
        // SAFETY: Param is read-only and doesn't allow mutable access to World.
        // It also matches the World this SystemState was created with.
        unsafe { self.get_unchecked_manual(world.as_unsafe_world_cell_readonly()) }
    }
```

This interface allows for the caller to provide a different world to `new` and to `get`, but I have to assume that would result in poor outcomes.

`new` stores `world_id`. I was wondering if there was a technical limitation preventing us from retrieving `World` by `world_id` from within `get` and eliminating the parameter?

**Comment by alice-i-cecile** [comment, marked as answer]
World access is required during system state initialization in order to generate all of the metadata needed to set up the system machinery.

World acccess is required when using system state in order to actually *get* the data from the world. We can't look it up by world_id, because that's just a dumb identifier: we need some sort of reference to the world data.

That said, we could improve boilerplate here by adding methods to initialize and get the system state immediately. The `SystemState` won't be cached anywhere (which is a nice micro-optimization), but it reduces the annoyance for cold work. `World::get_system_state` would probably be my preference; there's no point adding an immutable form since you always need write access on initialization to be able to register components and so on.

**Reply by MeoMix** [reply]
Okay! That all generally makes sense. I guess I thought that if we had a dumb identifier for World that it wouldn't be hard to look up the reference by id, but it sounds like that is technically challenging.

**Reply by alice-i-cecile** [reply]
Yeah, the question is "what are you looking it up from". We don't keep a collection of worlds on hand, so there's nothing to index into :)
